utils/threadpool/fair_threadpool.cpp

```diff
@@ -0,0 +1,295 @@
+    auto job = txnJobsList->begin();
+    while (job != txnJobsList->end())
+    {
+      if (job->id_ == id)
+      {
+        job = txnJobsList->erase(job); // update the job iter
+        if (txnJobsList->empty())
+        {
+          txn2JobsListMap_.erase(txnJobsMapPair.first);
+          delete txnJobsList;
+          break;
+          // There is no clean-up for PQ. It will happen later in threadFcn
+        }
+        continue; // go-on skiping job iter increment
+      }
+      ++job;
+    }
+  }
+}
+
+void FairThreadPool::threadFcn(const Priority preferredQueue)
+{
+  if (preferredQueue == EXTRA)
+    utils::setThreadName("Extra");
+  else
+    utils::setThreadName("Idle");
+  RunListT runList; // This is a vector to allow to grab multiple jobs
+  RescheduleVecType reschedule;
+  bool running = false;
+  bool rescheduleJob = false;
+
+  try
+  {
+    while (!_stop)
+    {
+      runList.clear(); // remove the job
+      boost::mutex::scoped_lock lk(mutex);
+
+      if (preferredQueue == EXTRA && stopExtra)
+      {
+        --extraThreads;
+        return;
+      }
+
+      if (weightedTxnsQueue_.empty())
+      {
+        newJob.wait(lk);
+        continue; // just go on w/o re-taking the lock
+      }
+
+      WeightedTxnT weightedTxn = weightedTxnsQueue_.top();
+      auto txnAndJobListPair = txn2JobsListMap_.find(weightedTxn.second);
+      // Looking for non-empty jobsList in a loop
+      // Waiting on cond_var if PQ is empty(no jobs in this thread pool)
+      while (txnAndJobListPair == txn2JobsListMap_.end() || txnAndJobListPair->second->empty())
+      {
+        if (txnAndJobListPair != txn2JobsListMap_.end()) // JobList is empty
+        {
+          ThreadPoolJobsList* txnJobsList = txnAndJobListPair->second;
+          delete txnJobsList;
+          txn2JobsListMap_.erase(txnAndJobListPair->first);
+        }
+        weightedTxnsQueue_.pop(); // remove the empty
+        if (weightedTxnsQueue_.empty())
+        {
+          break;
+        }
+        weightedTxn = weightedTxnsQueue_.top();
+        txnAndJobListPair = txn2JobsListMap_.find(weightedTxn.second);
+      }
+
+      if (weightedTxnsQueue_.empty())
+      {
+        newJob.wait(lk); // might need a lock here
+        continue;
+      }
+
+      // We have non-empty jobsList at this point.
+      // Remove the txn from a queue first to add it later
+      weightedTxnsQueue_.pop(); // remove the empty
+      TransactionIdxT txnIdx = txnAndJobListPair->first;
+      ThreadPoolJobsList* jobsList = txnAndJobListPair->second;
+      // Job& job = jobsList->front();
+      runList.push_back(jobsList->front());
+
+      jobsList->pop_front();
+      // Add the jobList back into the PQ adding some weight to it
+      // Current algo doesn't reduce total txn weight if the job is rescheduled.
+      if (!jobsList->empty())
+      {
+        weightedTxnsQueue_.push({weightedTxn.first + runList[0].weight_, txnIdx});
```

Comment:
Just for my education, why do we choose a new weight for transaction a sum of current weight + job weight?

Reply:
This allows to reduce a priority of the txn that owns the morsel(PP primitive job in the context) comparing with other txns so that FairThreadPool scheduling will pick another txn's morsel to run. In other words every time when txn runs a morsel in FTP it looses its priority to be scheduled next. However this algo doesn't affect single-query performance but adds some(presumably neglectable) overhead.

Reply:
Ok, thanks. In this case an important part is to calculate a weight for the specific job, the calculation algo is not covered in this part, as far as I see.

Reply:
I personally disagree this is an important part for the initial version.